README.md
# mentor.twolfson.com [![Build status](https://circleci.com/gh/twolfson/mentor.twolfson.com/tree/master.svg?style=svg)](https://circleci.com/gh/twolfson/mentor.twolfson.com/tree/master)

Website for twolfson's mentoring service

https://mentor.twolfson.com/

## Getting started
To run this repo locally, execute the following:

```bash
# Clone repo
git clone https://github.com/twolfson/mentor.twolfson.com

# Install our dependencies
npm install

# Run our server
npm start
# Server running at http://localhost:5000
```

Website will be accessible at: http://localhost:5000/index.html

## Documentation
We provide the following scripts/commands:

- `npm start`, start our local development server
  - Files will automatically be updated via Parcel.js' hot module replacement
  - This sometimes requires rebooting on file addition/deletion but is a slick alternative to LiveReload
- `npm run build`, builds files for production
- `foundry release <version>`, creates a new `git tag` version and updates `package.json` via `foundry`
  - Requires `foundry.cli` is installed
- `./deploy.sh <ssh-server>`, builds and deploys static files to an SSH server

## Unlicense
As of Aug 25 2020, Todd Wolfson has released this repository and its contents to the public domain.

It has been released under the [UNLICENSE][].

[UNLICENSE]: UNLICENSE


CHANGELOG.md
# mentor.twolfson.com
## 2021.1106.1219
- Bumped review count

## 2021.0513.0040
- Pinned parcel to version without "Invalid Version: undefined" error
- Updated available hours

## 2020.0909.1801
- Updated Open Graph to have correct type/url for pages
- Added canonical URL as future-proofing

## 2020.0908.2129
- Added per-language landing pages for more specific SEO/ads

## 2020.0908.1724
- Relocated server/views for easier partials and extension with less per-page noise in the future

## 2020.0905.2041
- Added sitemap.txt

## 2020.0830.1806
- Added 404 page
- Extended development server to closer match NGINX

## 2020.0830.0023
- Removed commas from Pug to make copy/paste with HTML much easier

## 2020.0830.0007
- Overall goal: Make it easier to track conversions by keeping a page on site somehow
- Added views layout and contact success page
- Added `package-lock.json` fix and script
- Corrected typo in reviews
- Added `_blank` for `mailto` links as well as a test

## 2020.0826.2212
- Rewrote server for test integration
- Added testing
- Added ESLint
- Added CircleCI
- Added IDs to all links for detailed tracking
- Added email link in Contact section

## 2020.0826.0335
- Added `xxs` styles for reviews
- Added SEO and analytics
- Visual tweaks for mobile

## 2020.0825.0829
Initial release


test/analytics.js
// Load in our dependencies
const cheerio = require('cheerio');
const { expect } = require('chai');
const { generateBundler } = require('../server/index.js');

// Define our tests
describe('Root/Contact page', function () {
  // DEV: If this becomes common, then memoize results in a common test file
  before(async function generateHTML() {
    this.timeout(30e3);
    this.bundler = generateBundler();
    this.bundle = await this.bundler.bundle();

    let rootBundle = Array.from(this.bundle.childBundles).find((bundle) => bundle.name.endsWith('/index.html'));
    this.htmlStr = rootBundle.entryAsset.generated.html;
    expect(this.htmlStr).to.be.a('string');
  });
  before(function generateCheerio() {
    this.$ = cheerio.load(this.htmlStr);
  });

  it('has an id on each `a` element', function () {
    let $aEls = this.$('a');
    expect($aEls.length).to.be.greaterThan(3);
    $aEls.each((i, aEl) => {
      let $aEl = this.$(aEl);
      expect($aEl.attr('id')).to.not.eq(undefined, $aEl + ' id');
    });
  });

  it('all elements have unique ids', function () {
    let $idEls = this.$('[id]');
    expect($idEls.length).to.be.greaterThan(3);
    let ids = $idEls.toArray().map((idEl) => this.$(idEl).attr('id'));
    let idSet = new Set(ids);
    expect(Array.from(idSet)).to.deep.equal(ids);
  });

  it('has an id on each `form` element', function () {
    let $formEls = this.$('form');
    // DEV: Using `eq(1)` until we can test multiple forms
    expect($formEls.length).to.eq(1);
    $formEls.each((i, formEl) => {
      let $formEl = this.$(formEl);
      expect($formEl.attr('id')).to.not.eq(undefined, $formEl + ' id');
    });
  });

  it('uses `_blank` on conversion links', function () {
    let $conversionLink = this.$('#contact__email-link');
    expect($conversionLink.length).to.eq(1);
    expect($conversionLink.attr('target')).to.eq('_blank');
  });
});


browser/js/bootcamp-alternative.js
// JS specific for bootcamp-alternative page (page-specific JS is an anti-pattern)
// When the page loads
// http://youmightnotneedjquery.com/#ready
// http://youmightnotneedjquery.com/#on
document.addEventListener('DOMContentLoaded', function handleReady () {
  // Define our compare change handler
  var lastSelection = 0;
  var selectEl = document.getElementById('compare-bootcamp-select');
  if (!selectEl) { throw new Error('Unable to find #compare-bootcamp-select'); }
  function handleCompareChange(evtOrNull) {
    // Resolve our current selection
    var currentSelection = selectEl.selectedIndex;

    // If our selection is unchanged, then do nothing
    if (lastSelection === currentSelection) {
      return;
    }

    // Otherwise, handle our change
    lastSelection = currentSelection;
    var compareInfoJSON = selectEl.options[currentSelection].getAttribute('data-compare-info');
    // DEV: We have to encode with quotes to avoid escaping to `&quot;` which is valid
    var compareInfo = JSON.parse(compareInfoJSON.replace(/'/g, '"'));
    document.getElementById('infrequent-cost').textContent = compareInfo.infrequent.cost;
    document.getElementById('infrequent-savings').textContent = compareInfo.infrequent.savings;
    document.getElementById('infrequent-sessions').textContent = compareInfo.infrequent.sessions;
    document.getElementById('regular-cost').textContent = compareInfo.regular.cost;
    document.getElementById('regular-savings').textContent = compareInfo.regular.savings;
    document.getElementById('regular-sessions').textContent = compareInfo.regular.sessions;
    document.getElementById('frequent-cost').textContent = compareInfo.frequent.cost;
    document.getElementById('frequent-savings').textContent = compareInfo.frequent.savings;
    document.getElementById('frequent-sessions').textContent = compareInfo.frequent.sessions;
  }

  // Run our compare change handler (in case the selection is off)
  handleCompareChange(null);

  // When our select option changes, run our compare change event
  selectEl.addEventListener('change', handleCompareChange);
});


server/index.js
// Heavily based on https://parceljs.org/api.html
// Load in our dependencies
const Bundler = require('parcel');

// Define our constants
const ENV_DEVELOPMENT = 'development';
const ENV_TEST = 'test';
const ENV_PRODUCTION = 'production';

// Fallback our environment
process.env.NODE_ENV = process.env.NODE_ENV || ENV_DEVELOPMENT;

// Set up Parcel configuration
const consolePaths = [
  '/',
  '/bootcamp-alternative.html',
  '/html-css.html',
  '/javascript.html',
  '/nodejs.html',
  '/python.html',
  '/ruby.html',
];
const entryFiles = [
  // DEV: When adding a new page, consider updating `sitemap.txt`
  __dirname + '/views/index.pug',
  __dirname + '/views/contact-success.pug',
  __dirname + '/views/404.pug',

  __dirname + '/views/bootcamp-alternative.pug',
  __dirname + '/views/html-css.pug',
  __dirname + '/views/javascript.pug',
  __dirname + '/views/nodejs.pug',
  __dirname + '/views/python.pug',
  __dirname + '/views/ruby.pug',

  __dirname + '/views/robots.txt',
  __dirname + '/views/sitemap.txt',
];
const options = {
  https: false,
  port: 5000,
};

// Extend options on a per-environment basis
if (process.env.NODE_ENV === ENV_DEVELOPMENT) {
  // No changes
} else if (process.env.NODE_ENV === ENV_TEST) {
  // DEV: This line seems to tell program to quit when done
  //   https://github.com/parcel-bundler/parcel/blob/v1.10.3/src/cli.js#L198
  options.production = true;
  options.logLevel = 2; // Errors and warnings only (no info)
} else if (process.env.NODE_ENV === ENV_PRODUCTION) {
  options.production = true;
}

// Generate and export a bundler for extension
// https://github.com/parcel-bundler/parcel/blob/v1.10.3/src/cli.js#L213-L228
exports.generateBundler = function () {
  return new Bundler(entryFiles, options);
};

// If this file is being run via the CLI, then run the following
// https://github.com/parcel-bundler/parcel/blob/v1.10.3/src/cli.js#L213-L228
/* eslint-disable no-console */
async function main() {
  let bundler = exports.generateBundler();
  if (process.env.NODE_ENV === ENV_DEVELOPMENT) {
    // Output our pages for developers
    console.info(`Pages:`);
    consolePaths.forEach((consolePath) => {
      console.info(`- http://localhost:${options.port}${consolePath}`);
    });

    // Start our server as per normal
    let server = await bundler.serve(options.port, options.https); // eslint-disable-line no-unused-vars
    // Skipping browser open logic

    // Override our server to behave more like NGINX
    // https://github.com/parcel-bundler/parcel/blob/v1.10.3/src/Server.js#L43-L72
    let _handleRequest = server._events.request;
    server._events.request = function (req, res, next) {
      // If someone is requesting `/index.html`, then redirect them to `/`
      // req.url = '/main.c5ebeee7.css';
      if (req.url.startsWith('/index.html')) {
        if (req.url.indexOf('?') !== -1) { throw new Error('Query string not supported in redirects yet'); }
        // DEV: 302 is temporary redirect (301 is permanent)
        // https://stackoverflow.com/a/4062281
        res.writeHead(302, {
          Location: '/'
        });
        res.end();
        return;
      }

      // If someone is requesting `/`, then serve it as `index.html`
      if (req.url === '/' || req.url === '/?') {
        req.url = '/index.html';
      }

      // DEV: Usually the callback/`next` is for more middleware but they seem to only use it for 404
      //   https://github.com/parcel-bundler/parcel/blob/v1.10.3/src/Server.js#L43-L111
      return _handleRequest.call(this, req, res, function custom404() {
        req.url = '/404.html';
        _handleRequest.call(this, req, res);
      });
    };
  } else if (process.env.NODE_ENV === ENV_PRODUCTION) {
    bundler.bundle();
  } else {
    throw new Error(`\`index.js\` being run in unexpected environment: ${process.env.NODE_ENV}`);
  }
}
if (require.main === module) {
  main();
}
